src/translate.rs

```diff
@@ -1081,56 +1158,119 @@ impl FromGlib<glib_sys::gboolean> for bool {
+/// Translate from a Glib type which can result in an undefined and/or invalid value.
+pub trait TryFromGlib<G>: Sized {
+    type Error;
+    fn try_from_glib(val: G) -> Result<Self, Self::Error>;
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be undefined.
+#[derive(Debug, PartialEq, Eq)]
+pub struct UndefinedError;
+
+impl fmt::Display for UndefinedError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "value is undefined")
+    }
+}
+
+impl std::error::Error for UndefinedError {}
+
+impl<G, T: TryFromGlib<G, Error = UndefinedError>> FromGlib<G> for Option<T> {
     #[inline]
-    fn from_glib(val: i32) -> Ordering {
-        val.cmp(&0)
+    fn from_glib(val: G) -> Option<T> {
+        T::try_from_glib(val).ok()
+    }
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be invalid.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidError;
+
+impl fmt::Display for InvalidError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "invalid conversion attempted")
     }
 }
 
-impl FromGlib<u32> for Option<char> {
+impl std::error::Error for InvalidError {}
+
+impl TryFromGlib<u32> for char {
+    type Error = InvalidError;
     #[inline]
-    fn from_glib(val: u32) -> Option<char> {
-        match val {
-            0 => None,
-            _ => char::from_u32(val),
-        }
+    fn try_from_glib(val: u32) -> Result<Self, InvalidError> {
+        char::from_u32(val).ok_or(InvalidError)
     }
 }
 
-impl FromGlib<i32> for Option<u32> {
+impl TryFromGlib<i32> for u32 {
+    type Error = InvalidError;
     #[inline]
-    fn from_glib(val: i32) -> Option<u32> {
-        if val >= 0 {
-            Some(val as u32)
-        } else {
-            None
-        }
+    fn try_from_glib(val: i32) -> Result<Self, InvalidError> {
+        u32::try_from(val).map_err(|_| InvalidError)
     }
 }
 
-impl FromGlib<i64> for Option<u64> {
+// FIXME should we implement this here?
+// Maybe we should force the caller to use try_from_glib instead.
+impl FromGlib<u32> for char {
     #[inline]
-    fn from_glib(val: i64) -> Option<u64> {
-        if val >= 0 {
-            Some(val as u64)
-        } else {
-            None
+    fn from_glib(val: u32) -> char {
+        char::try_from_glib(val).expect("Valid Unicode character expected")
+    }
+}
+
+impl TryFromGlib<i64> for u64 {
+    type Error = InvalidError;
+    #[inline]
+    fn try_from_glib(val: i64) -> Result<Self, InvalidError> {
+        u64::try_from(val).map_err(|_| InvalidError)
+    }
+}
+
+impl TryFromGlib<i32> for u64 {
+    type Error = InvalidError;
+    #[inline]
+    fn try_from_glib(val: i32) -> Result<Self, InvalidError> {
+        u64::try_from(val).map_err(|_| InvalidError)
+    }
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be undefined or invalid.
+#[derive(Debug, PartialEq, Eq)]
+pub enum UndefinedOrInvalidError {
+    Invalid,
+    Undefined,
```

Comment:
In the other direction you call it "none". That should be consistent. So either `None` here, other `UndefinedToGlib` for the trait. I prefer the "none" naming :)

Reply:
I'm going with None everywhere!